Make clippy happy without significant changes

nyurik opened the pull request:
* removed a number of unneeded lifetimes
* spelling
* `cargo fmt`
* added cargo fmt & clippy to the CI
* added Default and Debug to the `PathParser`
* lots of unneeded `&`
* discovered a bug in clippy in the process (link in code)
* made walk.rs faster (index access is always slower than iterative one because iterative doesn't need to do bounds check on each access
* fix deprecated constants
* use `DebugValidator` in the `PathCommands` (will use less memory now when Debug is not enabled)

nical reviewed (changes requested):
Thanks! Lots of great stuff in there, I'm amazed by the amount of typos you fixed all over the repository. Just a few things to change here and there before this is ready to land.

nyurik commented on .github/workflows/main.yml:
```diff
@@ -33,6 +33,18 @@ jobs:
+        uses: actions-rs/cargo@v1
+        with:
+          command: clippy
+          args: --all --all-targets --all-features -- -D warnings
```
sure, i will remove it for now (as it does fail on fmt/clippy).  I think [actions-rust-lang/rustfmt](https://github.com/marketplace/actions/check-formatting-of-rust-code-with-rustfmt) might work in the future - i will need to test it

nyurik commented:
@nical thank you for such a thorough review! I agree that Clippy is somewhat pushy when it comes to indexer vs iterator, and I agree it is a bit easier to read it in the current state, and usually I just give in to Clippy for one simple reason -- I tend not to have enough energy to go digging through the assembly to make sure Rust in each specific case has done "the right thing" - easier to succumb to the dark side (i'm a bit lazy that way) :smile: 

Regardless - I addressed all the above comments, let me know if anything else needs changing
